sled-storage/src/manager.rs

```diff
@@ -0,0 +1,969 @@
+            }
+        }
+    }
+
+    // Add a disk to `StorageResources` if it is new and return true if so
+    fn add_disk(&mut self, raw_disk: RawDisk) -> bool {
+        let disk = match raw_disk {
+            RawDisk::Real(_) => {
+                panic!(
+                    "Only synthetic disks can be used with `FakeStorageManager`"
+                );
+            }
+            RawDisk::Synthetic(synthetic_disk) => {
+                Disk::Synthetic(synthetic_disk)
+            }
+        };
+        self.resources.insert_fake_disk(disk)
+    }
+}
+
+/// The storage manager responsible for the state of the storage
+/// on a sled. The storage manager runs in its own task and is interacted
+/// with via the [`StorageHandle`].
+pub struct StorageManager {
+    log: Logger,
+    state: StorageManagerState,
+    rx: mpsc::Receiver<StorageRequest>,
+    resources: StorageResources,
+    queued_u2_drives: HashSet<RawDisk>,
+    key_requester: StorageKeyRequester,
+    resource_updates: watch::Sender<StorageResources>,
+}
+
+impl StorageManager {
+    pub fn new(
+        log: &Logger,
+        key_requester: StorageKeyRequester,
+    ) -> (StorageManager, StorageHandle) {
+        let (tx, rx) = mpsc::channel(QUEUE_SIZE);
+        let resources = StorageResources::default();
+        let (update_tx, update_rx) = watch::channel(resources.clone());
+        (
+            StorageManager {
+                log: log.new(o!("component" => "StorageManager")),
+                state: StorageManagerState::WaitingForKeyManager,
+                rx,
+                resources,
+                queued_u2_drives: HashSet::new(),
+                key_requester,
+                resource_updates: update_tx,
+            },
+            StorageHandle { tx, resource_updates: update_rx },
+        )
+    }
+
+    /// Run the main receive loop of the `StorageManager`
+    ///
+    /// This should be spawned into a tokio task
+    pub async fn run(&mut self) {
+        loop {
+            const QUEUED_DISK_RETRY_TIMEOUT: Duration = Duration::from_secs(10);
+            let mut interval = interval(QUEUED_DISK_RETRY_TIMEOUT);
+            interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+            tokio::select! {
+                res = self.step() => {
+                    if let Err(e) = res {
+                        warn!(self.log, "{e}");
+                    }
+                }
+                _ = interval.tick(),
+                    if self.state == StorageManagerState::QueuingDisks =>
+                {
+                    if self.add_queued_disks().await {
+                        let _ = self.resource_updates.send_replace(self.resources.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    /// Process the next event
+    ///
+    /// This is useful for testing/debugging
+    pub async fn step(&mut self) -> Result<(), Error> {
+        // The sending side should never disappear
```

Comment:
Hmm, I was thinking that I didn't ever want the task to shutdown. However, if there is no handle to it, it can't be communicated with and there's really nothing it's doing that's important while running. I'll go ahead and change this to shutdown in the case the sender is gone. I'll do it above as well, obviating your question about the comment.

Reply:
Fixed in c895657e1